### PR TITLE
enable lazydownload for CERN merge jobs

### PR DESCRIPTION
**Pull request by hufnagel**
All merge jobs running at CERN force lazy-download.

Originally forced the new mode Brian implemented, which works, but has currently such big performance problems that it can't be turned on. Still want to leave the code in (disabled with an "or False) because the code technically works and will be needed again if/when the performance problems get fixed.

### CODE REVIEW

**Comment by hufnagel** [comment]
@ticoann @ericvaandering would appreciate a review.

Without this any merge job run on CERN Cloud will potentially suffer from a factor 10 to 20 performance degradation...

File: src/python/WMCore/WMRuntime/Scripts/SetupCMSSWPset.py [hunk 1]
```diff
@@ -546,7 +546,28 @@ def __call__(self):
...
+
+            cmsswVersion = os.environ['CMSSW_VERSION']
+            if cmsswVersion.startswith("CMSSW_7_5") and False:
+                print "Using fastCloning/lazydownload"
+                self.process.add_(
+                    cms.Service("SiteLocalConfigService",
+                                overrideSourceCloneCacheHintDir=cms.untracked.string("lazy-download")
+                                )
+                    )
+            elif funcName == "merge":
```
Comment (ericvaandering) [inline src/python/WMCore/WMRuntime/Scripts/SetupCMSSWPset.py]: Why don't you just use an OR?

Comment (hufnagel) [inline src/python/WMCore/WMRuntime/Scripts/SetupCMSSWPset.py]: OR for what ? These are two different configuration changes !

Comment (hufnagel) [inline src/python/WMCore/WMRuntime/Scripts/SetupCMSSWPset.py]: One sets lazy-download, the sets "if the job uses fast cloning -> use lazy-download"

**Comment by ericvaandering** [comment]
I'd also suggest moving the deterimination of lazy download into a separate function. That main function si getting pretty long.

File: src/python/WMCore/WMRuntime/Scripts/SetupCMSSWPset.py [hunk 2]
```diff
@@ -546,7 +546,28 @@ def __call__(self):
...
+                print "Using lazydownload"
+                self.process.add_(
+                    cms.Service("SiteLocalConfigService",
+                                overrideSourceCacheHintDir=cms.untracked.string("lazy-download")
```
Comment (ericvaandering) [inline src/python/WMCore/WMRuntime/Scripts/SetupCMSSWPset.py]: Ah, I missed the absence of "Clone" in this line.

**Comment by ticoann** [comment]
I don't know anything about this, so I can review. :smile: 
But I have a question.
So lazy down load will only trigger in cern machine and only merge job?

File: src/python/WMCore/WMRuntime/Scripts/SetupCMSSWPset.py [hunk 3]
```diff
@@ -546,7 +546,28 @@ def __call__(self):
...
         self.fixupProcess()
 
-        # repack is only run at CERN and has a streaming format as input that should never use lazy-download
+        # for CERN:
+        #   enable lazy-download for fastCloning for all CMSSW_7_5 jobs (currently off)
+        #   enable lazy-download for all merge jobs at CERN
+        if socket.getfqdn().endswith("cern.ch"):
+
+            cmsswVersion = os.environ['CMSSW_VERSION']
+            if cmsswVersion.startswith("CMSSW_7_5") and False:
```
Comment (ticoann) [inline src/python/WMCore/WMRuntime/Scripts/SetupCMSSWPset.py]: Is this just reminder (to remove 'and False') for later since this always False? right?

Comment (ericvaandering) [inline src/python/WMCore/WMRuntime/Scripts/SetupCMSSWPset.py]: Right. The fix from Brian *should* land in 7.5 but it's not there yet.

**Comment by hufnagel** [comment]
Yes, the whole code block only triggers on CERN machines and only on merge jobs. Reason being that due to the weird CERN internal network layout (with two locations being 30ms apart), we cannot afford not to run lazy-download for merges but OTOH also cannot really afford to use lazy-download for all jobs due to how much disk space this uses on the local disk.

Brians patch would allow configuring this in the site-local-config and then every CMSSW job would automatically switch on lazy-download if the job was a merge job or did "merge-like" data handling.

So the second code block (that's currently disabled with the False) would test this behavior before we configure it in site-local-config.

@ericvaandering , I'll look at factoring this out into a separate function.

**Comment by hufnagel** [comment]
@ericvaandering , you like this better ? Tried a bit of code cleanup, unifying naming conventions, adding return statements to method, that sort of thing. Still not 100% internally consistent, but I didn't want to change too much either.

**Comment by hufnagel** [comment]
Tested in a Tier0 replay, repack and merge jobs behave in the correct way and all other jobs at least work as well (but I didn't check the logs for them).

**Comment by ericvaandering** [comment]
Yeah. Just two comments. First, I don't have an issue with starting a function name with underscore if it's not supposed to be called by anything but inside the class. This is kind of a convention we have and I don't think it's just us. 

Second, you modified where the lazy download fixups are handled to only be when a process is created. I hope that's what you want.

In any case, I am fine to merge this.

**Comment by hufnagel** [comment]
I didn't mind the convention, but I do mind a convention that is inconsistently applied. It doesn't make much sense to follow one convention for methods A, B, C and another for methods D, E, F. Plus from what I have seen of at least WMCore code, we rarely use the _method convention. It's still not completely consistent in this file, but as I said, I didn't want to change everything :-)

I moved the repack and merge modifications because for now they only apply to Config.DP created repack and merge processes. If Brian ever gets his patch working properly we'll have to revisit this anyways (best case is this method will disappear altogether and move to be configured from site-local-conf).

**Comment by hufnagel** [comment]
If there are no objections to the code, can we get this merged soon. I have a bunch of changes piled up and would like to build a new T0 release soon, preferably with this in it.

**Comment by hufnagel** [comment]
@ticoann ?

**Pull request merged**